ivy/functional/frontends/builtins/func_wrapper.py

```diff
@@ -0,0 +1,101 @@
+import functools
+from typing import Callable, Iterable
+from importlib import import_module
+
+import ivy
+
+
+def _to_ivy_array(x):
+    # if x is any frontend array including frontend list
+    # then extract the wrapped ivy array
+    if hasattr(x, "ivy_array"):
+        return x.ivy_array
+
+    # convert native arrays and lists to ivy arrays
+    elif isinstance(x, ivy.NativeArray):
+        return ivy.array(x)
+
+    return x
+
+
+def _infer_return_array(x: Iterable) -> Callable:
+    # get module path
+    module_path = x.__class__.__module__.split(".")
+
+    # if function's input is a scalar, list, or tuple
+    # convert to current backend's frontend array
+    if "builtins" in module_path:
+        cur_backend = ivy.current_backend_str()
+        module_str = cur_backend if len(cur_backend) != 0 else "numpy"
+
+    # in this case we're dealing with frontend array
+    # module's name is always at index 3 on the path
+    elif "ivy" in module_path:
+        module_str = module_path[3]
+
+    # native array, e.g. np.ndarray, torch.Tensor etc.
+    else:
+        module_str = module_path[0]
+        # replace jaxlib with jax to construct a valid path
+        module_str = "jax" if module_str == "jaxlib" else module_str
+
+    # import the module and get a corresponding frontend array
+    frontend_path = "ivy.functional.frontends." + module_str
+    module = import_module(frontend_path)
+    frontend_array = getattr(module, "_frontend_array")
+
+    return frontend_array
+
+
+def inputs_to_ivy_arrays(fn: Callable) -> Callable:
+    @functools.wraps(fn)
+    def _inputs_to_ivy_arrays_builtins(*args, **kwargs):
+        ivy_args = ivy.nested_map(args, _to_ivy_array, shallow=False)
+        ivy_kwargs = ivy.nested_map(kwargs, _to_ivy_array, shallow=False)
+
+        # array is the first argument given to a function
+        frontend_array = _infer_return_array(args[0]) if len(args) != 0 else None
+
+        return fn(*ivy_args, **ivy_kwargs), frontend_array
+
+    return _inputs_to_ivy_arrays_builtins
+
+
+def outputs_to_frontend_arrays(fn: Callable) -> Callable:
+    @functools.wraps(fn)
+    def _outputs_to_frontend_arrays(*args, **kwargs):
+        # update config for jax backend
+        if ivy.current_backend_str() == "jax":
+            import jax
+
+            jax.config.update("jax_enable_x64", True)
+
+        ret, frontend_array = fn(*args, **kwargs)
+
+        if isinstance(ret, (ivy.Array, ivy.NativeArray)):
+            return frontend_array(ret)
+        return ret
+
+    return _outputs_to_frontend_arrays
```

Comment:
Actually I think that built-in functions will always produce outputs in standard Python data types (scalars, lists, etc.), even when applied to iterables containing some kind of framework array/tensor. 
If I'm right, then we would not need `outputs_to_frontend_arrays` at all. We would have to make sure each builtin frontend function returns the type it's supposed to inside its implementation. For example, `min` should return a list, so at the end we would do `ivy.min(args).to_list()`.

Reply:
I was experimenting with this. Here's the results:
```
# Example for NumPy
>>> a = np.array([1, 2, 3, 4])
>>> type(abs(a))
numpy.ndarray
>>> type(all(a))
bool
>>> type(min(a))
numpy.int32
>>> type(max(a))
numpy.int32
>>> type(sum(a))
numpy.int32

# Example for torch
>>> a = torch.Tensor([1, 2, 3, 4])
>>> type(abs(a))
torch.Tensor
>>> type(all(a))
bool
>>> type(min(a))
torch.Tensor
>>> type(max(a))
torch.Tensor
>>> type(sum(a))
torch.Tensor
```
`min`, `max`, `sum` functions return 0-d array in this case, but it's still a `torch.Tensor`